Carthage  add transfer nonce

dobertRowneySr opened the pull request:
#### Changes
* add a `transfer_id` nonce to `PendingTransfer::transfer_params`
* add test to check that accepting a transfer with an invalid id fails
* add test to check that transfer id is increased after transfer is issued
* add tests to check that transfer status is correctly activated / disactivated
* refactor transfer tests



┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202132419573087/1202521734353275) by [Unito](https://www.unito.io)

L3pereira committed:
introduce_channel_state_bloat_bond_#3511

requested changes

Carthage T: Trait fix

L3pereira committed:
introduce_channel_state_bloat_bond_#3511

L3pereira committed:
introduce_channel_state_bloat_bond_#3511

requested changes

Carthage T: Trait fix

L3pereira committed:
Introduce video state bloat bond #3394

Introduce video state bloat bond #3394

dobertRowneySr committed:
test fixes

dobertRowneySr committed:
fix tests

dobertRowneySr committed:
fix tests

dobertRowneySr committed:
fix tests

dobertRowneySr committed:
add revenue_split_rate to token data

dobertRowneySr committed:
add revenue_split_rate to token data

dobertRowneySr committed:
add revenue_split_rate to token data

dobertRowneySr committed:
rebase

dobertRowneySr committed:
rebase

L3pereira committed:
remove_council_blog_pallet_and_proposals_#3662

remove_blog_proposals_#3662

remove_council_blog_pallet_and_proposals_#3662

remove_council_blog_typescript_pallet_and_proposals_#3662

remove_council_blog_typescript_pallet_and_proposals_#3662

requested changes + cargo fmt in types.rs

requests + types + metadata

chain metadata

dobertRowneySr committed:
remove categories from content

dobertRowneySr committed:
fix runtime

dobertRowneySr committed:
fix chainspec config

dobertRowneySr committed:
generate types library and metadata

L3pereira committed:
video state bloat bond Carthage T: Trait fix

Commit:
Update @polakdot/* deps, @joystream/types and configs for Substrate v3

Commit:
@joystream/types - augmentation by default (remove the need for tsconfig setup etc.)

dobertRowneySr committed:
add Transaction Id to runtime

dobertRowneySr committed:
add Transaction Id to runtime

dobertRowneySr committed:
fix all tests

dobertRowneySr committed:
add test for error with wrong transfer id

dobertRowneySr committed:
add test for re-setting status

Commit:
Types - update docs and add some utility functions

Commit:
Update api-scripts (new `@polkadot/api` and `@joystream/types`)

dobertRowneySr committed:
simplify and fix complete_payment helper and refactor tests

dobertRowneySr committed:
aremove channel caterogy id from Genesis Config

Commit:
Update chain metadata and augmentations

Commit:
Make buy-membership api-script more reusable

dobertRowneySr committed:
fix clippy

Commit:
Migration scripts - remove redundant "paths" from tsconfig.json

Commit:
@joystream/types: Import augmented registry from `augment/registry.ts` to allow skiping generic type args for `createType`

dobertRowneySr committed:
add test case & refactor

dobertRowneySr committed:
test refactoring

dobertRowneySr committed:
add test for activating transfer

Commit:
Update chain metadata

Commit:
Remove redundant ChannelCategory type

Commit:
@joystream/types - expose aliases for runtime primitives

Commit:
`createType` - `unknown` return type if type not found

Commit:
@joystream/types - simplify CreateInterface for BTreeMap/BTreeSet

Commit:
createType - remove generic Codec arg (T), rely only on TypeName (TN)

Commit:
CreateInterface - support for `Raw`

Commit:
CreateInterface: AnyU8a for Bytes | Raw

dobertRowneySr committed:
fix clippy

dobertRowneySr committed:
add blank line

dobertRowneySr committed:
generate types

dobertRowneySr committed:
generate types

dobertRowneySr committed:
update types and metadata

dobertRowneySr committed:
set next_sale_id starting value to 0

dobertRowneySr committed:
set next_sale_id starting value to 0

dobertRowneySr committed:
set next_sale_id starting value to 0

dobertRowneySr committed:
add test with failing destination member

dobertRowneySr committed:
rename allocation_amount -> revenue_amount
So the terminology is:
- allocation amount = what creator allocates for a revenue split
- revenue amount = what is actually provided to shareholders

dobertRowneySr committed:
disallow issuing a token with 0 revenue split rate

dobertRowneySr committed:
add test for lefover revenue split allocation amount

dobertRowneySr committed:
add test for lefover revenue split allocation amount

dobertRowneySr committed:
generate metadata & types

dobertRowneySr committed:
generate metadata & types

dobertRowneySr committed:
yarn format

mnaamani committed:
at least one nomination if initial nominators setup at genesis

L3pereira committed:
types generation and channel duplication fix

types generation

types generation and channel duplication fix

L3pereira committed:
types generation

L3pereira committed:
types generation

dobertRowneySr committed:
swap names allocation <-> revenue

dobertRowneySr committed:
swap names allocation <-> revenue

dobertRowneySr committed:
swap revenue <-> allocation in tests & traits

dobertRowneySr committed:
update chain meta data

dobertRowneySr committed:
send fund to controller account after rev. split

dobertRowneySr committed:
add test for leftover funds

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -2356,25 +2361,27 @@ decl_module! {
             channel_id: T::ChannelId,
             actor: ContentActor<T::CuratorGroupId, T::CuratorId, T::MemberId>,
-            new_transfer_status: ChannelTransferStatus<T::MemberId, T::CuratorGroupId, BalanceOf<T>>
+            candidate_status: ChannelTransferStatus<T::MemberId, T::CuratorGroupId, BalanceOf<T>, T::TransferId>
```
The sender here can provide any `transfer_id`, which doesn't really make any difference, it's just a useless parameter. I think this should be refactored, so that only parameters which actually affect the result can be provided.

This can be achieved, for example, by introducing a new `ChannelTransferStatusParams` type

dobertRowneySr replied:
How about I split `update_channel_transfer_status` into:
- `activate_channel_transfer`
- `cancel_channel_transfer`

Lezek123 replied:
Sounds good, `update_channel_transfer_status` always seemed a little awkward to me.
I'd just suggest `initialize_channel_transfer` as a name for the first function.

dobertRowneySr committed:
fix naming inconsistencies

dobertRowneySr committed:
refactor build genesis config

Lezek123 commented on runtime-modules/content/src/tests/transfers.rs:
```diff
@@ -232,158 +196,166 @@ fn accept_transfer_status_fails_with_non_channel_owner() {
-        let price = INITIAL_BALANCE + 1; // higher than initial balance
+        ContentTest::with_member_channel().setup();
+        let balance_pre = Balances::<Test>::usable_balance(&DEFAULT_MEMBER_ACCOUNT_ID);
+        let price = balance_pre.saturating_add(DEFAULT_CHANNEL_TRANSFER_PRICE);
```
I'm not sure why would we need to care about `DEFAULT_MEMBER_ACCOUNT_ID` balance here

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -74,6 +74,9 @@ pub trait Config:
     type OpenAuctionId: NumericIdentifier;
 
+    /// Type of identifier for TransferId
+    type TransferId: NumericIdentifier;
```
Missing implementation in `runtime/src/lib.rs`

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -156,6 +159,8 @@ decl_storage! {
 
         pub NextVideoId get(fn next_video_id) config(): T::VideoId;
 
+        pub NextTransferId get(fn next_transfer_id) config(): T::TransferId;
```
Missing initialization in `src/chain_spec/content_config.rs`

dobertRowneySr committed:
add extra checks on actor balances for edge case

Lezek123 reviewed (changes requested):
There's a missing test case to make sure that the nonce gets incremented if you do `update_channel_transfer_status(PendingTransfer)` => `update_channel_transfer_status(NoActiveTransfer)` =>`update_channel_transfer_status(PendingTransfer)`

dobertRowneySr committed:
add extra checks on normal case with royalty

dobertRowneySr committed:
cargo format

Lezek123 commented:
Metadata & types also need to be updated

dobertRowneySr committed:
fix naming inconsistency 2

dobertRowneySr committed:
cargo format

mnaamani committed:
Update node/src/chain_spec/mod.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

mnaamani committed:
only build arm64 and amd64 docker images

mnaamani committed:
drop armv7 platform job for joystream-node docker

dobertRowneySr committed:
fix clippy

dobertRowneySr committed:
generate metadata

dobertRowneySr committed:
add test to verify transfer nonce is correcly updated

dobertRowneySr committed:
fix Runtime Content::Trait & chainspec

dobertRowneySr committed:
Merge branch 'carthage' into carthage__add_transfer_nonce

dobertRowneySr committed:
add guard for cancelling channel transfer

I did not add any new permission I just added a guard
with an appropriate name

dobertRowneySr committed:
add CancelChannelTransfer event

dobertRowneySr committed:
fix clippy

dobertRowneySr committed:
add fixture for cancel channel transfer

dobertRowneySr committed:
add tests for cancel transfer

dobertRowneySr committed:
Merge branch 'carthage' into carthage__add_transfer_nonce

dobertRowneySr committed:
fix transfer tests & fixtures

dobertRowneySr committed:
rename: TransferParameters -> TransferCommitmentParameters

dobertRowneySr committed:
refactor and add test for init_channel_transfer

dobertRowneySr committed:
all tests passing

dobertRowneySr committed:
generate types & metadata

dobertRowneySr committed:
rename tests cases

dobertRowneySr committed:
add permission test on initialize channel transfer

dobertRowneySr committed:
add test for testing transfer permissions

dobertRowneySr committed:
cargo format

dobertRowneySr commented:
I have split the `update_channel_transfer` status in `initialize_channel_transfer` and `cancel_channel_transfer`.
I have also added extra tests for the new extr. introduce + test for testing that transfer permissions works as expected

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -372,7 +372,7 @@ decl_module! {
 
         /// Add curator to curator group under given `curator_group_id`
-        #[weight = 10_000_000] // TODO: adjust weight
+        #[weight = 10_000_0] // adjust weight
```
I think this was changed by mistake

Lezek123 commented on runtime-modules/content/src/tests/fixtures.rs:
```diff
@@ -5535,7 +5455,7 @@ pub fn run_all_fixtures_with_contexts(
             .with_actor(actor)
             .call_and_assert(expected_err.clone());
-        UpdateChannelTransferStatusFixture::default()
+        InitializeChannelTransferFixture::default()
```
You should also add the new `CancelChannelTransferFixture` here

Lezek123 commented on runtime-modules/content/src/tests/transfers.rs:
```diff
@@ -8,78 +8,333 @@ use sp_core::sp_std::iter::FromIterator;
+}
+
+#[test]
+fn initialize_channel_transfer_ok_when_issued_by_curator() {
```
So for tests like:
- `initialize_channel_transfer_ok_when_issued_by_lead`,
- `initialize_channel_transfer_ok_when_issued_by_curator`,
- `initialize_channel_transfer_ok_when_issued_by_collaborator`

There are already test cases in `tests/channels.rs` under:
```rust
///////////////////////////////////////////////////////////////////
/////// Channel agent / owner permissions - TransferChannel ///////
///////////////////////////////////////////////////////////////////
```

Namely:
- `successful_transfer_by_curator_agent`
- `successful_channel_transfer_by_lead`
- `successful_channel_transfer_by_collaborator`

Lezek123 commented on runtime-modules/content/src/tests/transfers.rs:
```diff
@@ -8,78 +8,333 @@ use sp_core::sp_std::iter::FromIterator;
+}
+
+#[test]
+fn initialize_channel_transfer_fails_with_lead_and_non_curator_channel() {
```
Test cases like:
- `initialize_channel_transfer_fails_with_lead_and_non_curator_channel`
- `initialize_channel_transfer_fails_with_curator_non_authorized_channel`
- `initialize_channel_transfer_fails_with_member_actor`
- `initialize_channel_transfer_fails_with_non_channel_owner`

Will already be covered by:
- `unsuccessful_curator_channel_actions_with_invalid_context`
- `unsuccessful_member_channel_actions_with_invalid_context`

I added those to normalize invalid-context tests across all related content extrinsics, since creating them manually was a very repetitive and error-prone task

Lezek123 commented on runtime-modules/content/src/tests/transfers.rs:
```diff
@@ -232,290 +453,327 @@ fn accept_transfer_status_fails_with_non_channel_owner() {
-            .with_price(price)
-            .call_and_assert(Ok(()));
+#[test]
+fn cancel_channel_transfer_fails_with_non_member_owner() {
```
As mentioned in the previous comment, cases like:
- `cancel_channel_transfer_fails_with_non_member_owner`
- `cancel_channel_transfer_fails_with_lead_and_member_channel`
- `cancel_channel_transfer_fails_with_invalid_curator`

Should be covered by:
- `unsuccessful_curator_channel_actions_with_invalid_context` and
- `unsuccessful_member_channel_actions_with_invalid_context`

(you just need to update `run_all_fixtures_with_contexts` accordingly to one of the previous comments)

Lezek123 commented on runtime-modules/content/src/tests/transfers.rs:
```diff
@@ -232,290 +453,327 @@ fn accept_transfer_status_fails_with_non_channel_owner() {
 #[test]
-fn update_transfer_status_blocked_during_upcoming_revenue_split() {
-    pub const SPLIT_STARTING_BLOCK: u64 = 10;
+fn cancel_channel_transfer_fails_with_collaborator_not_having_permissions() {
```
Tests like:
- `cancel_channel_transfer_fails_with_collaborator_not_having_permissions`
- `cancel_channel_transfer_ok_with_member_owner`
- `cancel_channel_transfer_ok_with_lead`
- `cancel_channel_transfer_ok_with_collaborator_having_permissions`
- `cancel_channel_transfer_ok_with_curator_having_permissions`

are usually part of `tests/channel.rs`, because they're related to channel agent permissions and I thought it would be more clear to have all such tests in one place.

So, for example, we have:
```rust
///////////////////////////////////////////////////////////////////
/////// Channel agent / owner permissions - TransferChannel ///////
///////////////////////////////////////////////////////////////////

// Curator channels

fn unsuccessful_channel_transfer_by_curator_agent_without_permissions() { /* ... */ }
fn successful_transfer_by_curator_agent() { /* ... */ }
fn successful_channel_transfer_by_lead() { /* ... */ }

// Member channels

fn unsuccessful_channel_transfer_by_collaborator_without_permissions() { /* ... */ }
fn successful_channel_transfer_by_collaborator() { /* ... */ }
fn successful_channel_transfer_by_owner_member() { /* ... */ }
```

And I think you can add similar ones for `cancel_channel_transfer` there also (and remove them from here)

dobertRowneySr replied:
I have added tests to the `tests/channel.rs`

dobertRowneySr committed:
fix transfer test (those related to permissions)

dobertRowneySr committed:
Merge branch 'carthage' into carthage__add_transfer_nonce

dobertRowneySr committed:
generate types and metadata

dobertRowneySr committed:
Merge branch 'carthage' into carthage__add_transfer_nonce

dobertRowneySr committed:
generate types and metadata

dobertRowneySr committed:
tests clippy fix

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -371,7 +376,7 @@ decl_module! {
 
         /// Add curator to curator group under given `curator_group_id`
-        #[weight = 10_000_000] // TODO: adjust weight
+        #[weight = 10_000_000] // adjust weight
```
```suggestion
        #[weight = 10_000_000] // TODO: adjust weight
```

Lezek123 reviewed (changes requested):
I think you missed one of the previous comments: https://github.com/Joystream/joystream/pull/3972#discussion_r913919850 (those test need to be added in `tests/channels.rs`)

Commit:
Update runtime-modules/content/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
add cancel transfer permission related tests

dobertRowneySr committed:
Merge branch 'carthage' into carthage__add_transfer_nonce

dobertRowneySr committed:
generate metadata

Commit:
Fix channel agent permissions test cases for initialize/cancel channel transfer

Lezek123 reviewed (changes requested):
Maybe I wasn't clear enough about what I meant, but for channel agent permissions tests there is a certain structure, where each extrinsic (`X`) has 6 tests that cover all the agent/owner permissions related cases:
```rust
/////////////////////////////////////////////////////
/////// Channel agent / owner permissions - X ///////
/////////////////////////////////////////////////////

// Curator channels

// Curator with unsufficient permissions case
fn unsuccessful_X_by_curator_agent_without_permissions() { /* ... */ }

// Curator with correct permissions case
fn successful_X_by_curator_agent() { /* ... */ }

// Lead case
fn successful_X_by_lead() { /* ... */ }

// Member channels

// Collaborator with unsufficient permissions case
fn unsuccessful_X_by_collaborator_without_permissions() { /* ... */ }

// Collaborator with correct permissions case
fn successful_X_by_collaborator() { /* ... */ }

// Owner member case
fn successful_X_by_owner_member() { /* ... */}
```

I wanted you to add the tests that match this structure for `cancel_channel_transfer` extrinsic.

I created a PR to your branch that ilustrates this: https://github.com/ignazio-bovo/joystream/pull/6 (if you agree with the changes you can just merge this PR)

Commit:
Merge pull request #6 from Lezek123/carthage__add_transfer_nonce

Fix channel agent permissions test cases for initialize/cancel channel transfer

Lezek123 reviewed (approved):
Great improvements, LGTM
